File: src/main/java/hjUDPproxy.java
/* hjUDPproxy, 20/Mar/18
 *
 * This is a very simple (transparent) UDP proxy
 * The proxy can listening on a remote source (server) UDP sender
 * and transparently forward received datagram packets in the
 * delivering endpoint
 *
 * Possible Remote listening endpoints:
 *    Unicast IP address and port: configurable in the file config.properties
 *    Multicast IP address and port: configurable in the code
 *
 * Possible local listening endpoints:
 *    Unicast IP address and port
 *    Multicast IP address and port
 *       Both configurable in the file config.properties
 */

import coins.CoinsRepository;
import kotlin.Triple;
import org.jetbrains.annotations.NotNull;
import rtstp.RTSTPNegotiatorClient;
import sadkdp.auth.AuthClient;
import sadkdp.dto.TicketCredentialsDto;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.security.KeyStore;
import java.security.Security;
import java.util.Arrays;
import java.util.Properties;
import java.util.Set;
import java.util.stream.Collectors;

class hjUDPproxy {
    public static void main(String[] args) throws Exception {
        InputStream inputStream = null;
        try {
            inputStream = new FileInputStream("config/proxy/config.properties");
        } catch (FileNotFoundException e) {
            System.err.println("Configuration file not found!");
            System.exit(1);
        }
        Security.addProvider(new org.bouncycastle.jce.provider.BouncyCastleProvider());
        Properties properties = new Properties();
        properties.load(inputStream);
        String remote = properties.getProperty("remote");
        String destinations = properties.getProperty("localdelivery");

        String signal = properties.getProperty("signal");

        SocketAddress inSocketAddress = parseSocketAddress(remote);
        Set<SocketAddress> outSocketAddressSet = Arrays.stream(destinations.split(","))
                .map(hjUDPproxy::parseSocketAddress)
                .collect(Collectors.toSet());

        System.out.print("Starting proxy server");
        AuthClient auth = new AuthClient(
                new CoinsRepository(),
                inSocketAddress,
                parseSocketAddress(signal),
                getKeyStoreFromFile("PKCS12", "config/proxy/proxy.p12", "password")
        );
        Triple<TicketCredentialsDto.Payload, byte[], byte[]> streamInfo = auth.getStreamInfo("user", "password", "proxyBoxId", "coinId", "cars");
        DatagramSocket inSocket = new RTSTPNegotiatorClient(streamInfo, parseSocketAddress(properties.getProperty("streaming"))).negotiate();
        DatagramSocket outSocket = new DatagramSocket();
        byte[] buffer = new byte[4 * 1024];


        while (true) {
            DatagramPacket inPacket = new DatagramPacket(buffer, buffer.length);
            inSocket.receive(inPacket); // if remote is unicast

            System.out.print("*");
            for (SocketAddress outSocketAddress : outSocketAddressSet) {
                outSocket.send(new DatagramPacket(inPacket.getData(), inPacket.getLength(), outSocketAddress));
            }
        }
    }

    private static InetSocketAddress parseSocketAddress(String socketAddress) {
        String[] split = socketAddress.split(":");
        String host = split[0];
        int port = Integer.parseInt(split[1]);
        return new InetSocketAddress(host, port);
    }

    @NotNull
    private static KeyStore getKeyStoreFromFile(String type, String fileName, String password) throws Exception {
        KeyStore keyStore = KeyStore.getInstance(type);
        FileInputStream stream = new FileInputStream(fileName);
        keyStore.load(stream, password.toCharArray());
        return keyStore;
    }
}

File: ProxyServer/src/hjUDPproxy.kt
import java.io.FileInputStream
import kotlin.Throws
import kotlin.jvm.JvmStatic
import java.io.InputStream
import java.lang.Exception
import java.util.Properties
import java.net.SocketAddress
import java.util.Arrays
import java.net.InetSocketAddress
import java.util.stream.Collectors
import java.net.DatagramSocket
import java.net.DatagramPacket

internal object hjUDPproxy {
    @Throws(Exception::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val inputStream: InputStream = FileInputStream("config.properties")
        if (inputStream == null) {
            System.err.println("Configuration file not found!")
            System.exit(1)
        }
        val properties = Properties()
        properties.load(inputStream)
        val remote = properties.getProperty("remote")
        val destinations = properties.getProperty("localdelivery")
        val inSocketAddress: SocketAddress = parseSocketAddress(remote)
        val outSocketAddressSet =
            Arrays.stream(destinations.split(",".toRegex()).toTypedArray()).map { s: String -> parseSocketAddress(s) }
                .collect(Collectors.toSet())
        val inSocket = DatagramSocket(inSocketAddress)
        val outSocket = DatagramSocket()
        val buffer = ByteArray(4 * 1024)
        while (true) {
            val inPacket = DatagramPacket(buffer, buffer.size)
            inSocket.receive(inPacket) // if remote is unicast
            print("*")
            for (outSocketAddress in outSocketAddressSet) {
                outSocket.send(DatagramPacket(buffer, inPacket.length, outSocketAddress))
            }
        }
    }

    private fun parseSocketAddress(socketAddress: String): InetSocketAddress {
        val split = socketAddress.split(":".toRegex()).toTypedArray()
        val host = split[0]
        val port = split[1].toInt()
        return InetSocketAddress(host, port)
    }
}